doc/architectural/CBMC-developer-guide.md

```diff
@@ -90,101 +275,242 @@ The `–help` option gives instructions for use and the
 format of the description files.
 
 
-\section general-info General Information
 
-First off, read the \ref cbmc-user-manual "CBMC User Manual". It describes
-how to get, build and use CBMC. This document covers the
-internals of the system and how to get started on development.
 
-## Documentation
+\section section-data-structures-core-structures-and-ast Data structures: core structures and AST
 
-Apart from the (user-orientated) CBMC user manual and this document, most
-of the rest of the documentation is inline in the code as `doxygen` and
-some comments. A man page for CBMC, goto-cc and goto-instrument is
-contained in the `doc/` directory and gives some options for these
-tools. All of these could be improved and patches are very welcome. In
-some cases the algorithms used are described in the relevant papers.
+## Strings: dstringt, the string_container and the ID_* ##
+Within cbmc, strings are represented using irep_idt. By default this is
+typedefed to dstringt, which stores a string as an index into a large
+static table of strings. This makes it easy to compare if two irep_idts
+are equal (just compare the index) and it makes it efficient to store
+many copies of the same string. The static list of strings is populated
+from irep_ids.def, so for example the fourth entry in irep_ids.def is
+“IREP_ID_ONE(type)”, so the string “type” has index 3. You can refer to
+this irep_idt as ID_type. The other kind of line you see is
+“IREP_ID_TWO(C_source_location, #source_location)”, which means the
+irep_idt for the string “#source_location” can be referred to as
+ID_C_source_location. The “C” is for comment, which will be explained
+in the next section.
 
-## Architecture
+See documentation at \ref irep_idt.
 
-This section provides a graphical overview of how CBMC fits together.
-CBMC takes C code or a goto-binary as input and tries to emit traces
-of executions that lead to crashes or undefined behaviour. The diagram
-below shows the intermediate steps in this process.
+## Irept: a 4-triple (data, named-sub, comments-sub, sub) ##
```

Comment:
Suggest put `sub` after `data` as the most-used part

Reply:
We should probably be consistent with the order they are listed in the documentation for `irept` (cf #2618), which is this order. It's also the order that they're listed in in `irep.h`.